de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/SarosSessionManager.java

```diff
@@ -485,8 +497,56 @@ public void invite(Collection<JID> jidsToInvite, String description) {
    * @param projectResourcesMapping
    */
   @Override
-  public void addResourcesToSession(Map<IProject, List<IResource>> projectResourcesMapping) {
+  public synchronized void addResourcesToSession(
+      Map<IProject, List<IResource>> projectResourcesMapping) {
+    if (projectResourcesMapping == null) {
+      return;
+    }
+
+    /*
+     * To prevent multiple concurrent negotiations per user. 1. Collect all
+     * new mappings, 2. If active negotiations are running wait till they
+     * finish (collect new mappings in the meantime), 3. Create one
+     * negotiation with all collected resources.
+     */
+
+    nextProjectNegotiation.add(projectResourcesMapping);
 
+    if (nextProjectNegotiationWorker != null && nextProjectNegotiationWorker.isAlive()) {
+      return;
+    } else if (currentProjectNegotiations.isEmpty()) {
+      /* shortcut to direct handling */
+      startNextProjectNegotiation();
+      return;
+    }
+
+    /* else create a worker thread */
+    Runnable worker =
+        new Runnable() {
+          @Override
+          public void run() {
+            synchronized (nextProjectNegotiation) {
+              while (!currentProjectNegotiations.isEmpty()) {
+                try {
+                  nextProjectNegotiation.wait();
+                } catch (InterruptedException e) {
+                  Thread.currentThread().interrupt();
+                  return;
+                }
+              }
+            }
+
+            startNextProjectNegotiation();
+          }
+        };
+    nextProjectNegotiationWorker = ThreadUtils.runSafeAsync(log, worker);
+  }
+
+  /**
+   * This method handles new project negotiations for already invited user (not the first in the
+   * process of inviting to the session).
+   */
+  private synchronized void startNextProjectNegotiation() {
```

Comment:
Please pull the synchronized down to line 603 so it matches the former synchronized block. It should be clear what section must be synchronized. Synchronizing the whole method is not very helpful for understanding.